### DISCUSSION TITLE
Passing Python function to different Rust thread

### DISCUSSION THREAD

**Discussion by marcpabst** [PyO3/pyo3#3788] [Questions] [answered]
Hey all, I'm trying to adapt a Rust API for Python using PyO3 but I've hit a bit of a roadblock with threading. 

I have an existing Rust API that looks something like this:

```rust
pub fn run_experiment<F>(&mut self, window_options: &WindowOptions, experiment_fn: F) -> ()
where
    F: FnOnce(Window) -> Result<(), SomeErrorType>
        + 'static
        + Send,
```

This function uses `winit` to create a new EventLoop that blocks the main thread, but crucially runs the `experiment_fn` function in its own separate thread. How would I make this API available through PyO3? I'd like to mimic the same signature in Python, i.e., doing something like this (wrapping WindowOptions in a wrapper type):

```rust
fn run_experiment(
    &mut self,
    window_options: &PyWindowOptions,
    experiment_fn: &PyFunction,
) {
    // create rust FnOnce(Window) -> Result<(), SomeErrorType> + 'static + Send,
    let rust_experiment_fn = move |window: Window| -> Result<
        (),
        SomeErrorType,
    > {
        experiment_fn.call0(); // call without args for now
        Ok(())
    };

    // run the experiment
    self.0.run_experiment(
        &window_options.0,
        rust_experiment_fn,
    );
}
```

But that gives me a "`UnsafeCell<PyObject>` cannot be shared between threads safely" error, as `&PyFunction` is not `Send`. Any pointers to what would be the best approach to implement this?

**Comment by birkenfeld** [comment, marked as answer]
You'll have to use GIL-independent objects, i.e. `experiment_fn: Py<PyFunction>`. In `rust_experiment_fn`, you then have to acquire the GIL before you can call `experiment_fn`.

**Reply by marcpabst** [reply]
Thanks! This seemed to work at first, but now `Python::with_gil` just seems to block forever:

```rust
  // THIS DOES NOT WORK
  fn run_experiment(
        &mut self,
        py: Python,
        window_options: &PyWindowOptions,
        experiment_fn: Py<PyFunction>,
    ) {
     let rust_experiment_fn = move |window: Window| -> Result<
            (),
            errors::PsychophysicsError,
        > {
            Python::with_gil(|py| -> PyResult<()> {
                let pywin = PyWindow(window);
                experiment_fn.call1(py, (pywin,)).unwrap();
                Ok(())
            })
            .unwrap();
            Ok(())
        };

        self.0.run_experiment(&window_options.0, rust_experiment_fn);
    }
}
```

However, if I just tell PyO3 that I have the GIL, it works:

```rust
    // THIS WORKS
    fn run_experiment(
        &mut self,
        py: Python,
        window_options: &PyWindowOptions,
        experiment_fn: Py<PyFunction>,
    ) {

        let rust_experiment_fn = move |window: Window| -> Result<
            (),
            errors::PsychophysicsError,
        > {
            unsafe {
                let py = Python::assume_gil_acquired();
                // Python::with_gil(
                //     |py| -> PyResult<()> {
                let pywin = PyWindow(window);
                experiment_fn.call1(py, (pywin,)).unwrap();
                //         Ok(())
                //     },
                // )
                // .unwrap();
            }
            Ok(())
        };

        self.0.run_experiment(&window_options.0, rust_experiment_fn);
    }
}
```

Any idea why I can't seem to get the GIL with `Python::with_gil`? (please ignore the non-handling of errors for now)

**Reply by adamreichold** [reply]
> This seemed to work at first, but now Python::with_gil just seems to block forever:

The problem is that the main thread already has the GIL when your `run_experiment` `#[pymethod]` is called. You need to release it so that the background can successfully acquire it to avoid the deadlock, e.g.

```rust
  fn run_experiment(
        &mut self,
        py: Python,
        window_options: &PyWindowOptions,
        experiment_fn: Py<PyFunction>,
    ) {
     let rust_experiment_fn = move |window: Window| -> Result<
            (),
            errors::PsychophysicsError,
        > {
            Python::with_gil(|py| -> PyResult<()> {
                let pywin = PyWindow(window);
                experiment_fn.call1(py, (pywin,)).unwrap();
                Ok(())
            })
            .unwrap();
            Ok(())
        };

        py.allow_threads(|| {
            self.0.run_experiment(&window_options.0, rust_experiment_fn);
        });
    }
}
```

**Reply by marcpabst** [reply]
The problem in this case is that `allow_threads` requires `f` to be `Send`, which it isn't because `self` is not `Send` (and doesn't have to be). But looking at `allow_threads`, it doesn't really seem to do anything special... unfortunately, `SuspendGIL` is not `pub`, so I can't just use that.

Would using my unsafe variant be fine, given that the Rust function is blocking and I know that I will never run any Python code concurrently?

**Reply by adamreichold** [reply]
> Would using my unsafe variant be fine, given that the Rust function is blocking and I know that I will never run any Python code concurrently?

I don't think so because acquiring the GIL also sets some thread-local state, i.e. the main thread has the GIL and not the background thread which cannot run Python code even if the main thread is blocked and cannot do so either.

I suspect you will have to encapsulate things so that `self.0` into the closure passed to `allow_threads` somehow.

**Reply by marcpabst** [reply]
Reading the docs was actually really helpful :)

this does the trick:
```rust
let mut self_wrapper = SendWrapper::new(self);
py.allow_threads(move || {
    self_wrapper
        .0
        .run_experiment(&window_options.0, rust_experiment_fn)
});
```

**Reply by adamreichold** [reply]
This is abusing a soundness loophole that we are currently trying to fix.

**Reply by davidhewitt** [reply]
(#3646 - although it's worth noting that `unsafe { py.allow_threads().local_unconstrained() }.with(|| self.0.run_experiment(...))` will probably work, adding merit to that solution.)

**Reply by marcpabst** [reply]
> This is abusing a soundness loophole that we are currently trying to fix.

One could claim that PyO3 is currently abusing `Send` :) I think I will stick with this for now and move to the new solution once that's stabilised.
